python/spark_sklearn/keyed_models.py

```diff
@@ -0,0 +1,407 @@
+        if len(keyCols) == 0:
+            raise ValueError("keyCols should not be empty")
+        if "estimator" in keyCols + [xCol, yCol]:
+            raise ValueError("keyCols should not contain a column named \"estimator\"")
+
+        for paramName, paramSpec in KeyedEstimator.paramSpecs.items():
+            setattr(self, paramName, Param(Params._dummy(), paramName, paramSpec["doc"]))
+        super(KeyedEstimator, self).__init__()
+        self._setDefault(**{paramName: paramSpec["default"]
+                            for paramName, paramSpec in KeyedEstimator.paramSpecs.items()
+                            if "default" in paramSpec})
+        kwargs = KeyedEstimator._inferredParams(self.__init__._input_kwargs)
+        self._set(**kwargs)
+
+        self._verifyEstimatorType()
+
+    @staticmethod
+    def _inferredParams(inputParams):
+        if "yCol" in inputParams:
+            inputParams["estimatorType"] = "predictor"
+        return inputParams
+
+    def _verifyEstimatorType(self):
+        estimatorType = self.getOrDefault("estimatorType")
+        estimator = self.getOrDefault("sklearnEstimator")
+
+        if not hasattr(estimator, "fit"):
+            raise ValueError("sklearnEstimator missing fit()")
+
+        if estimatorType == "transformer":
+            if not hasattr(estimator, "transform"):
+                raise ValueError("estimatorType assumed to be transformer, but " +
+                                 "sklearnEstimator is missing transform()")
+        elif estimatorType == "predictor":
+            if not hasattr(estimator, "predict"):
+                raise ValueError("estimatorType assumed to be predictorr, but " +
+                                 "sklearnEstimator is missing predict()")
+        else:
+            raise ValueError("estimatorType {} is unknown".format(estimatorType))
+
+    def _fit(self, dataset):
+        keyCols = self.getOrDefault("keyCols")
+        xCol = self.getOrDefault("xCol")
+        yCol = self.getOrDefault("yCol")
+        isLabelled = yCol is not None
+        assert isLabelled == (self.getOrDefault("estimatorType") == "predictor")
+        _validateXCol(dataset.schema, xCol)
+
+        cols = keyCols[:]
+        cols.append(xCol)
+        if isLabelled: cols.append(yCol)
+
+        oneDimensional = _isOneDimensional(dataset.schema, xCol)
+        projected = dataset.select(*cols) # also verifies all cols are present
+        outputSchema = StructType().add("estimator", _SparkSklearnEstimatorUDT.sqlType())
+        grouped = projected.groupBy(*keyCols)
+        estimator = self.getOrDefault("sklearnEstimator")
+
+        # Potential optimization: broadcast estimator
+
+        import pandas as pd
+        def fitEstimator(_, pandasDF):
+            X = _prepareXCol(pandasDF[xCol], oneDimensional)
+            y = pandasDF[yCol].values if isLabelled else None
+            pandasDF = None
```

Comment:
RESPONSE REQUESTED: what do you mean? I'm saying I don't need pandasDF anymore. `del`? Avoid the optimization altogether (also below too, then?)